### PR TITLE
gh-60580: Fix a wrong type of `ctypes.wintypes.BYTE`

### PR DESCRIPTION
**Pull request by arhadthedev**
This PR is created from a .diff file attached to the issue, by Anatoly Techtonik (@techtonik on GitHub).

* Issue: gh-60580

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2022-09-26-21-18-47.gh-issue-60580.0hBgde.rst
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Library/2022-09-26-21-18-47.gh-issue-60580.0hBgde.rst]: This is failing the build, maybe you need `:data:`?

Comment (arhadthedev) [reply Misc/NEWS.d/next/Library/2022-09-26-21-18-47.gh-issue-60580.0hBgde.rst]: Thank you, fixed.

**Comment by arhadthedev** [comment]
Does such a trivial change (addition of a single character) requires the CLA?

**Comment by arhadthedev** [comment]
@techtonik Would you mind to sign the CLA to get your PR acepted please? See a message of a bot above for details.

**Comment by AlexWaygood** [comment]
> Does a trivial change like this (addition of a single character) warrant the CLA?

Unfortunately it is impossible for a core dev to merge a PR without the CLA being signed, as the  CLA check is a "required" status check.

**Comment by arhadthedev** [comment]
What would you recommend to do? The change is way too trivial to give a CLA-covered alternative. Also it seems impossible to call a user who never contributed to `python/cpython`before.

**Comment by AlexWaygood** [comment]
> What would you recommend to do?

Squash the PR so that GitHub no longer recognises the first commit as being authored by @techtonik (thus evading the CLA bot), but make sure to give them credit in the commit message and PR description. Since this is, as you say, a small PR, I think that should be fine.

**Comment by arhadthedev** [comment]
@abalkin would you mind merging this PR as a ctypes expert?

**Comment by arhadthedev** [comment]
> Can someone provide a use case where this change makes a difference?

The only case I've found inside CPython is [`struct`](https://docs.python.org/3/library/struct.html) that maps `c_byte` to `b` and `c_ubyte` to `B`: https://github.com/python/cpython/blob/dfad678d7024ab86d265d84ed45999e031a03691/Lib/ctypes/__init__.py#L224-L225 https://github.com/python/cpython/blob/dfad678d7024ab86d265d84ed45999e031a03691/Lib/ctypes/__init__.py#L231-L232 https://github.com/python/cpython/blob/dfad678d7024ab86d265d84ed45999e031a03691/Modules/_struct.c#L808-L811

However, `n*_byte()` is the same as `n*_cbyte()`.

**Comment by arhadthedev** [comment]
It seems that external shim generation tools that might rely on the layout string will stay compatible.

**Comment by abalkin** [comment]
Curiously, [the same question](https://github.com/python/cpython/issues/60580#issuecomment-1093598930) was asked when this issue was first raised:

> All cbytes tests passed after the change was made but I've no idea as to the impact of this.

A test was then [suggested](https://github.com/python/cpython/issues/60580#issuecomment-1257250164) that should be sensitive to the change.  We need to add a unit test before this change can be merged.

**Comment by abalkin** [comment]
We can just add a simple signedness test.

Before the patch:

```python
>>> ctypes.wintypes.BYTE(200).value < 128
True
```

after the patch:

```python
>>> ctypes.wintypes.BYTE(200).value < 128
False
```

Since `ctypes.c_byte` is defined as signed, I would not be surprised if there is code in the wild relying on the wrong signedness of the `ctypes.wintypes.BYTE` alias.  We may need to make this change a little more visible possibly adding a note to the "porting to" section of the relevant  "What’s New In Python" page.

**Comment by abalkin** [comment]
@Yhg1s - please advise if this bug fix is appropriate for 3.12 and what documentation it will require in "What’s New In Python"?

TL;DR:  This PR changes `ctypes.wintypes.BYTE` alias from signed to unsigned char type to match the eponymous Window API typedef.

File: Lib/test/test_ctypes/test_wintypes.py
Comment (abalkin) [inline Lib/test/test_ctypes/test_wintypes.py]: I had to check the spelling myself, but the right word appear to be ["signedness"](https://en.wikipedia.org/wiki/Signedness).

File: Lib/test/test_ctypes/test_wintypes.py
Comment (abalkin) [inline Lib/test/test_ctypes/test_wintypes.py]: Please leave a space after `#` in comments.

File: Lib/test/test_ctypes/test_wintypes.py
Comment (arhadthedev) [reply Lib/test/test_ctypes/test_wintypes.py]: Fixed.

File: Lib/test/test_ctypes/test_wintypes.py
Comment (abalkin) [inline Lib/test/test_ctypes/test_wintypes.py]: Does this pass?

```python
>>> ctypes.wintypes.WORD(200).value
200
```

Comment (abalkin) [reply Lib/test/test_ctypes/test_wintypes.py]: Let's make signedness test a WinTypesTest method and implement it like this:

```python
    def assertIsSigned(self, ctype):
        self.assertLess(ctype(-1).value, 0)
```

Comment (arhadthedev) [reply Lib/test/test_ctypes/test_wintypes.py]: Thanks for the advice. I also got rid of the second loop over unsigned types to avoid further confusion between similar looking code blocks.

File: Lib/test/test_ctypes/test_wintypes.py
Comment (abalkin) [inline Lib/test/test_ctypes/test_wintypes.py]: This should be

```python
    self.assertIsSigned(type)
```

This is Python, not C++ - implicit "this" is not a feature. :)

File: Lib/test/test_ctypes/test_wintypes.py
Comment (abalkin) [inline Lib/test/test_ctypes/test_wintypes.py]: Please rename type to ctype to avoid conflict with the built-in method.  While not an error here, many IDEs and linters will flag it as such.

File: Lib/test/test_ctypes/test_wintypes.py
Comment (arhadthedev) [reply Lib/test/test_ctypes/test_wintypes.py]: Thank you, fixed again. At least, `python -m test test_ctypes.test_wintypes` now gives `Tests result: SUCCESS`.

It seems to be a bad idea to build a lean Qt 5 while working on the PR :D (I threw in a bunch of `--no-feature-*` into its `configure` and now meticulously remove them one by one after each hour-long build attempt because Qt components need most of the features and fail the build, reporting a single missing one each time).

So my apologises for such mistakes. I hope I've made no extra mistake.

**Comment by Yhg1s** [comment]
> @Yhg1s - please advise if this bug fix is appropriate for 3.12 and what documentation it will require in "What’s New In Python"?

Yes, it's appropriate for 3.12. (We're not in beta phase yet, so any fix appropriate for any release is appropriate for 3.12.) I don't think it's worth mentioning in What's New, since it's just a minor bugfix.

**Review by abalkin — APPROVED** [review]

**Comment by CristiFati** [comment]
Hmm, I wonder how anyone (including me - as I used *wintypes.BYTE* a lot in the past) missed this. I was just about to open a *PR*, as I'm working with [\[Ms.Learn\]: SYSTEM\_POWER\_STATUS structure (winbase.h)](https://learn.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-system_power_status).

Regarding the tests: [\[SO\]: Maximum and minimum value of C types integers from Python (@CristiFati's answer)](https://stackoverflow.com/a/52485502/4788546).

**Pull request merged**
